workflow-bot-app/src/main/java/com/symphony/bdk/workflow/engine/camunda/CamundaExecutor.java

```diff
@@ -147,27 +148,39 @@ public CamundaActivityExecutorContext(DelegateExecution execution, T activity, E
     }
 
     @Override
-    public void setOutputVariable(String name, Object value) {
-      Map<String, Object> innerMap = Collections.singletonMap(name, value);
-      Map<String, Object> outerMap = Collections.singletonMap(ActivityExecutorContext.OUTPUTS, innerMap);
+    public void setOutputVariables(Map<String, Object> variables) {
+      Map<String, Object> innerMap = new HashMap<>(variables);
       String activityId = getActivity().getId();
 
-      // value might not implement serializable or be a collection with non-serializable items, so we use JSON if needed
-      Object outerMapVar;
-      Object valueVar;
-      if (value instanceof Serializable && !(value instanceof Collection)) {
-        outerMapVar = outerMap;
-        valueVar = value;
-      } else {
-        outerMapVar =
-            Variables.objectValue(outerMap).serializationDataFormat(Variables.SerializationDataFormats.JSON).create();
-        valueVar =
-            Variables.objectValue(value).serializationDataFormat(Variables.SerializationDataFormats.JSON).create();
+      Map<String, Object> outer = Map.of(ActivityExecutorContext.OUTPUTS, innerMap);
+      ObjectValue objectValue = Variables.objectValue(outer)
+          .serializationDataFormat(Variables.SerializationDataFormats.JSON)
+          .create();
+
+      // flatten outputs for message correlation
+      Map<String, Object> flattenOutputs = new HashMap<>();
+
+      for (Map.Entry<String, Object> entry : innerMap.entrySet()) {
+        // value might not implement serializable or be a collection with non-serializable items, so we use JSON if needed
+        if (entry.getValue() instanceof Serializable && !(entry.getValue() instanceof Collection)) {
+          flattenOutputs.put(entry.getKey(), entry.getValue());
+        } else {
+          flattenOutputs.put(entry.getKey(), Variables.objectValue(entry.getValue())
+              .serializationDataFormat(Variables.SerializationDataFormats.JSON)
+              .create());
+        }
       }
 
-      execution.setVariable(activityId, outerMapVar);
-      // flatten it too for message correlation
-      execution.setVariable(String.format("%s.%s.%s", activityId, ActivityExecutorContext.OUTPUTS, name), valueVar);
+      execution.setVariable(activityId, objectValue);
+      flattenOutputs.forEach((key, value) -> execution.setVariable(
+          String.format("%s.%s.%s", activityId, ActivityExecutorContext.OUTPUTS, key), value));
+    }
+
+    @Override
+    public void setOutputVariable(String name, Object value) {
+      Map<String, Object> singletonMap = new HashMap<>();
+      singletonMap.put(name, value);
+      this.setOutputVariables(singletonMap);
```

Comment:
Map.of() doesn't support entries with null value. Using  Hashmap this way avoid us to check if value is not null.
In some activities (update message, send message..) the output value cannot be null and then Map.of() could be used in this case, but I prefer to do it this way to avoid any surprises.

Reply:
and do we want to allow users to add null values?

Reply:
I would say yes as BDK service could return null (I don't have an example in my mind honestly).. It can also happens in custom activities depending on which service/api are used.